common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -1217,6 +1264,20 @@ public boolean toByte(IntWrapper intWrapper) {
 
   @Override
   public String toString() {
+    byte[] bytes = getBytes();
+    // Optimization for ASCII characters: use deprecated string API which
+    // skips charset encoder and simply casts each byte into a char.
+    if (isAscii(bytes)) {
+      return new String(bytes, 0);
+    } else {
+      return new String(bytes, StandardCharsets.UTF_8);
+    }
+  }
+
+  /**
+   * Pre-SPARK-27841 version of toString(), retained for benchmarking purposes.
+   */
+  public String toStringSlow() {
```

Comment:
Can this be package private?

Reply:
I don't think so: the benchmarks that I added are in a different package (because it's hard to depend on our benchmarks package from the `unsafe` package), so this needs to be public.

Reply:
Can we use `@VisibleForTesting` annotation?